server/odc-service/src/main/java/com/oceanbase/odc/service/iam/ResourceRoleService.java

```diff
@@ -75,35 +74,28 @@ public List<UserResourceRole> saveAll(List<UserResourceRole> userResourceRoleLis
         if (CollectionUtils.isEmpty(userResourceRoleList)) {
             return Collections.emptyList();
         }
-        List<UserResourceRole> userResourceRoles = new ArrayList<>();
-        Map<ResourceRoleName, ResourceRoleEntity> resourceRoleEntities =
-                resourceRoleRepository.findAll().stream().collect(Collectors.toMap(
-                        ResourceRoleEntity::getRoleName, v -> v, (existingValue, newValue) -> newValue));
-
-        List<UserResourceRoleEntity> entities = userResourceRoleList.stream().map(i -> {
-            ResourceRoleEntity resourceRoleEntity = resourceRoleEntities.getOrDefault(i.getResourceRole(), null);
-            if (resourceRoleEntity == null) {
-                throw new UnexpectedException("resource role not found, role=" + i.getResourceRole());
+        List<UserResourceRoleEntity> userResourceRoleEntityList = new ArrayList<>();
+        Map<ResourceType, Map<ResourceRoleName, ResourceRoleEntity>> resourceRoleMap = resourceRoleRepository.findAll()
+                .stream().collect(Collectors.groupingBy(ResourceRoleEntity::getResourceType,
+                        Collectors.toMap(ResourceRoleEntity::getRoleName, v -> v, (v1, v2) -> v2)));
+        userResourceRoleList.forEach(e -> {
+            Map<ResourceRoleName, ResourceRoleEntity> resourceRoleEntityMap = resourceRoleMap.get(e.getResourceType());
+            if (Objects.isNull(resourceRoleEntityMap)) {
+                throw new UnexpectedException("resource type has no role, type=" + e.getResourceType());
+            }
+            ResourceRoleEntity resourceRoleEntity = resourceRoleEntityMap.get(e.getResourceRole());
+            if (Objects.isNull(resourceRoleEntity)) {
+                throw new UnexpectedException("resource role not found, role=" + e.getResourceRole());
             }
             UserResourceRoleEntity entity = new UserResourceRoleEntity();
-            entity.setResourceId(i.getResourceId());
-            entity.setUserId(i.getUserId());
+            entity.setResourceId(e.getResourceId());
+            entity.setUserId(e.getUserId());
             entity.setResourceRoleId(resourceRoleEntity.getId());
             entity.setOrganizationId(authenticationFacade.currentOrganizationId());
-            userResourceRoles.add(fromEntity(entity, resourceRoleEntity));
-            return entity;
-        }).collect(Collectors.toList());
-
-        String batchInsert =
-                "insert into `iam_user_resource_role` (`user_id`, `resource_id`, `resource_role_id`, `organization_id`) VALUES (?, ?, ?, ?) on duplicate key update `id` = `id`";
-        JdbcTemplate jdbcTemplate = new JdbcTemplate(dataSource);
-        jdbcTemplate.batchUpdate(batchInsert, entities, entities.size(), (ps, entity) -> {
```

Comment:
Why do we delete the batch insert logic? You may extends `OdcJpaRepository` to make it more readable.

Reply:
Yes, done!